docs/design/storagenode-graceful-exit-protocol.md

```diff
@@ -0,0 +1,215 @@
+
+The satellites should set the `finished_at` on success, and respond with a `DeletePiece` message. Otherwise set the `failed_at` and `failure_status_code` for reprocessing.
+
+The satellite should respond with a `ExitCompleted` message when all pieces have finished processing. 
+The storage node should store the completion receipt, stop transfer processing, and remove all remaining pieces for the satellite.
+
+When the storage node has failed too many transfers or has sent incorrect data, the satellite will send `ExitFailed`. This indicates that the process has ended ungracefully.
+
+### Transferring a Piece
+
+The satellite picks a piece from the transfer queue and verifies whether it still needs to be transferred. During gathering or the Graceful Exit process the segment may have been changed or deleted.
+
+The satellite will send a `TransferPiece`, which contains sufficient information to upload the piece to a replacement node.
+
+The storage node will start a new piece upload to the replacement node similar to uplink. It will use `uplink/piecestore.Upload`. Once uploaded, the storage node will verify the piece hash sent by the replacement node corresponds to the piece hash stored in the database.
+
+#### Verifying transfer
+
+The storage node sends the piece hash, order limit, original "uploader" signature, and piece hash signed by the replacement node to the satellite for confirmation.
+The satellite verifies that the original piece hash matches the replacement piece hash, and verifies the order limits signature. On success, the satellite will update segment / pointer information. If verification fails, the satellite will send `ExitFailed`.
+
+![Transfer Sequence](./images/storagenode-graceful-exit-transfer-sequence.svg)
+
+#### Updating the Segment / Pointer
+
+When the upload has been verified the satellite will use `CompareAndSwap` to only switch the appropriate node without changing others. During the upload there may have been other things that happened:
+
+- Segment / pointer was reuploaded. We'll ignore piece transferred for the exit.
+- Segment / pointer pieces were changed, in that case we load the pointer and retry `CompareAndSwap`. When the storage node at that piece number has changed, we'll ignore the piece transferred for the exit, because one of them needs to be discarded.
+- Segment / pointer was deleted or is being deleted. We'll ignore the piece transferred for the exit.
+
+After changing the segment / pointer, we'll delete the piece from transfer queue, update progress, and send a `DeletePiece` to the exiting node.
+
+#### Handling transfer failure
+
+If a storage node report a failure to the satellite if it fails to upload to the replacement node, or the replacement node returns a bad piece hash. The satellite will track the failures and retry the transfer to another node.
```

Comment:
Hmm, not really... My understanding is that the satellite will send `ExitFailed` if the exiting nodes send the satellite info that looks like cheating. Here we are talking the transfer may fail because the replacement node returns a hash that the exiting node does not agree to.

Reply:
Here, the exiting node should report it as "transfer failed". The other part is about exiting node sending a piece hash which is invalid.